File: hist_name.js
#!/usr/bin/env node
// hist_name.js
const { program } = require('commander');
const fs = require('fs');
const chalk = require('chalk');
const crypto = require('crypto');

const NAMES = {
    viking: {
        male: ['Эйнар','Рагнар','Бьёрн','Свен','Ульф','Харальд','Олаф','Эрик','Торгейр','Гуннар','Сигурд','Хакон','Кнуд','Орм','Кетиль','Асбьёрн','Торстейн','Эгиль','Гудмунд','Хельги'],
        female: ['Астрид','Брюнхильд','Гудрун','Сигриди','Гунхильд','Хельга','Ингеборг','Тордис','Тора','Альвильд','Герда','Фрейдис','Раннвейг','Сигрид','Эрна','Грета','Кристин','Хильда','Руна','Сванхильд'],
        surnames: ['Бьёрнсон','Эйнарсон','Харальдсон','Свенсон','Рагнарсон','Ульфсон','Олафсон','Гуннарсон','Сигурдсон','Торгерсон','Кетильсон','Асбьёрнсон','Торстейнсон','Эгильсон','Гудмундсон'],
        titles: ['Берсерк','Скальд','Молот','Боевой Топор','Кровавый','Северный','Грозный','Мудрый']
    },
    roman: {
        male: ['Гай','Луций','Марк','Квинт','Публий','Секст','Тит','Авл','Децим','Гней','Сервий','Нумерий','Август','Тиберий','Клавдий','Нерон','Траян','Адриан','Антоний','Юлий'],
        female: ['Юлия','Ливия','Валерия','Клавдия','Цецилия','Теренция','Октавия','Друзилла','Агриппина','Мессалина','Антония','Корнелия','Помпея','Вителлия','Германика','Фавстина','Луцилла','Криспина','Сабина','Вероника'],
        surnames: ['Цезарь','Август','Нерон','Клавдий','Тиберий','Траян','Адриан','Антоний','Юлий','Германик','Брут','Сципион','Катон','Цицерон','Сенека','Аврелий','Флавий','Константин'],
        titles: ['Император','Консул','Сенатор','Легат','Претор','Эдил','Цензор','Диктатор']
    },
    medieval: {
        male: ['Уильям','Генри','Ричард','Роберт','Эдуард','Джон','Томас','Джеффри','Роджер','Хью','Саймон','Уолтер','Гилберт','Реджинальд','Балдуин','Алан','Брайан','Филип','Джеймс','Дэвид'],
        female: ['Элеонора','Матильда','Алиенора','Агнес','Изабелла','Беатрис','Сибилла','Маргарет','Кэтрин','Джоан','Алиса','Сесилия','Ида','Мария','Клеменция','Амелия','Филиппа','Гвендолин','Аделаида','Энн'],
        surnames: ['Нормандский','Анжуйский','Ланкастерский','Йоркский','Плантагенет','Уэссекский','Глостерский','Кентский','Суффолкский','Ричмондский','Пемброк','Мортимер','Грей','Стюарт','Говард'],
        titles: ['Король','Герцог','Граф','Барон','Рыцарь','Лорд','Епископ','Маркиз']
    },
    greek: {
        male: ['Александр','Аристотель','Демокрит','Эпикур','Гераклит','Платон','Сократ','Пифагор','Гомер','Софокл','Еврипид','Эсхил','Перикл','Леонид','Фемистокл','Агамемнон','Ахилл','Одиссей','Гектор','Тесей'],
        female: ['Афина','Гера','Афродита','Артемида','Деметра','Персефона','Гестия','Ника','Ирида','Афродита','Кассандра','Андромеда','Елена','Пенелопа','Антигона','Электра','Ифигения','Медея','Сапфо','Гипполита'],
        surnames: ['Афинский','Спартанский','Коринфский','Фиванский','Македонский','Фессалийский','Критский','Родосский','Милетский','Эфесский'],
        titles: ['Царь','Тиран','Архонт','Стратег','Философ','Поэт','Воин','Герой']
    },
    slavic: {
        male: ['Владимир','Святослав','Ярослав','Всеволод','Изяслав','Мстислав','Олег','Игорь','Святополк','Вячеслав','Ярополк','Глеб','Борис','Андрей','Александр','Дмитрий','Михаил','Пётр','Иван','Василий'],
        female: ['Ольга','Ярославна','Владислава','Людмила','Мирослава','Светлана','Евдокия','Борислава','Велеслава','Добромира','Вера','Надежда','Любовь','Мария','Анна','Екатерина','Анастасия','Татьяна','Елена','Ирина'],
        surnames: ['Рюрикович','Владимирович','Святославич','Ярославич','Всеволодович','Изяславич','Мстиславич','Олегович','Игоревич','Святополкович'],
        titles: ['Князь','Воевода','Боярин','Дружинник','Мудрый','Грозный','Святой','Великий']
    },
    celtic: {
        male: ['Айдан','Брендан','Коннор','Деклан','Эйлин','Финн','Гэвин','Иан','Киран','Лайам','Мэлколм','Ниалл','Она','Патрик','Рори','Шон','Тиадг','Улисс','Эмон','Брайан'],
        female: ['Айслин','Бриджит','Кэтлин','Диана','Эйлиш','Финола','Грэйн','Иона','Кира','Лианна','Майв','Ниам','Она','Рона','Сиана','Тара','Уна','Фиона','Сиобан','Морриган'],
        surnames: ['МакКауд','О\'Брайен','О\'Салливан','МакКарти','О\'Доннелл','МакДонах','О\'Нейл','МакГрат','О\'Ши','МакМахон','О\'Коннор','МакДауэлл','О\'Киф','МакГиннис','О\'Флаэрти'],
        titles: ['Король','Вождь','Друид','Воин','Мудрец','Певец','Охотник','Мастер']
    }
};

class HistoricalNameGenerator {
    constructor(options) {
        this.culture = options.culture || 'viking';
        this.gender = options.gender || 'male';
        this.count = options.count || 1;
        this.surname = options.surname || false;
        this.title = options.title || false;
        this.seed = options.seed || null;
        this.color = options.color || process.stdout.isTTY;
        if (this.seed !== null) {
            let s = this.seed;
            this._random = () => {
                s = (s * 9301 + 49297) % 233280;
                return s / 233280;
            };
        } else {
            this._random = () => crypto.randomInt(0, 1e9) / 1e9;
        }
        this.cultureData = NAMES[this.culture] || NAMES.viking;
    }

    generateName() {
        const names = this.cultureData[this.gender] || this.cultureData.male;
        let name = names[Math.floor(this._random() * names.length)];
        if (this.surname && this.cultureData.surnames) {
            const sname = this.cultureData.surnames[Math.floor(this._random() * this.cultureData.surnames.length)];
            name += ' ' + sname;
        }
        if (this.title && this.cultureData.titles) {
            const title = this.cultureData.titles[Math.floor(this._random() * this.cultureData.titles.length)];
            name += ' ' + title;
        }
        return name;
    }

    generate() {
        return Array.from({ length: this.count }, () => this.generateName());
    }

    print(names) {
        const cultureLabels = { viking:'викинг', roman:'римское', medieval:'средневековое', greek:'греческое', slavic:'славянское', celtic:'кельтское' };
        const cultureLabel = cultureLabels[this.culture] || this.culture;
        const genderLabel = this.gender === 'male' ? 'мужские' : 'женские';
        if (this.color) {
            console.log(chalk.cyan(`🏛️ Исторические имена (${cultureLabel}, ${genderLabel}):`));
            names.forEach((n, i) => console.log(`${i+1}. ${chalk.green(n)}`));
        } else {
            console.log(`Исторические имена (${cultureLabel}, ${genderLabel}):`);
            names.forEach((n, i) => console.log(`${i+1}. ${n}`));
        }
    }

    exportJson(names, filename) {
        fs.writeFileSync(filename, JSON.stringify({ culture: this.culture, gender: this.gender, names }, null, 2));
    }

    exportCsv(names, filename) {
        const content = 'name\n' + names.join('\n');
        fs.writeFileSync(filename, content);
    }

    exportText(names, filename) {
        fs.writeFileSync(filename, names.join('\n'));
    }
}

program
    .option('-c, --culture <culture>', 'culture: viking, roman, medieval, greek, slavic, celtic', 'viking')
    .option('-g, --gender <gender>', 'gender: male, female', 'male')
    .option('--count <number>', 'Number of names', parseInt, 1)
    .option('--surname', 'Add surname')
    .option('--title', 'Add title')
    .option('--seed <number>', 'Seed', parseInt)
    .option('--output <file>', 'Output file')
    .option('--format <format>', 'Format: text, json, csv')
    .parse(process.argv);

const opts = program.opts();
const gen = new HistoricalNameGenerator(opts);
const names = gen.generate();

if (opts.output) {
    const ext = opts.output.split('.').pop().toLowerCase();
    const fmt = opts.format || (ext === 'json' ? 'json' : ext === 'csv' ? 'csv' : 'text');
    if (fmt === 'json') gen.exportJson(names, opts.output);
    else if (fmt === 'csv') gen.exportCsv(names, opts.output);
    else gen.exportText(names, opts.output);
    console.log(`Результат сохранён в ${opts.output}`);
} else {
    gen.print(names);
}

File: hist_name.py

#!/usr/bin/env python3
# hist_name.py
import argparse
import json
import csv
import random
import sys
from colorama import init, Fore, Style

init(autoreset=True)

# Базы данных имён по культурам и полам
NAMES = {
    "viking": {
        "male": [
            "Эйнар", "Рагнар", "Бьёрн", "Свен", "Ульф", "Харальд", "Олаф", "Эрик",
            "Торгейр", "Гуннар", "Сигурд", "Хакон", "Кнуд", "Орм", "Кетиль",
            "Асбьёрн", "Торстейн", "Эгиль", "Гудмунд", "Хельги"
        ],
        "female": [
            "Астрид", "Брюнхильд", "Гудрун", "Сигриди", "Гунхильд", "Хельга",
            "Ингеборг", "Тордис", "Тора", "Альвильд", "Герда", "Фрейдис",
            "Раннвейг", "Сигрид", "Эрна", "Грета", "Кристин", "Хильда", "Руна", "Сванхильд"
        ],
        "surnames": [
            "Бьёрнсон", "Эйнарсон", "Харальдсон", "Свенсон", "Рагнарсон",
            "Ульфсон", "Олафсон", "Гуннарсон", "Сигурдсон", "Торгерсон",
            "Кетильсон", "Асбьёрнсон", "Торстейнсон", "Эгильсон", "Гудмундсон"
        ],
        "titles": ["Берсерк", "Скальд", "Молот", "Боевой Топор", "Кровавый", "Северный", "Грозный", "Мудрый"]
    },
    "roman": {
        "male": [
            "Гай", "Луций", "Марк", "Квинт", "Публий", "Секст", "Тит", "Авл",
            "Децим", "Гней", "Сервий", "Нумерий", "Август", "Тиберий", "Клавдий",
            "Нерон", "Траян", "Адриан", "Антоний", "Юлий"
        ],
        "female": [
            "Юлия", "Ливия", "Валерия", "Клавдия", "Цецилия", "Теренция",
            "Октавия", "Друзилла", "Агриппина", "Мессалина", "Антония", "Корнелия",
            "Помпея", "Вителлия", "Германика", "Фавстина", "Луцилла", "Криспина",
            "Сабина", "Вероника"
        ],
        "surnames": [
            "Цезарь", "Август", "Нерон", "Клавдий", "Тиберий", "Траян",
            "Адриан", "Антоний", "Юлий", "Германик", "Брут", "Сципион",
            "Катон", "Цицерон", "Сенека", "Аврелий", "Флавий", "Константин"
        ],
        "titles": ["Император", "Консул", "Сенатор", "Легат", "Претор", "Эдил", "Цензор", "Диктатор"]
    },
    "medieval": {
        "male": [
            "Уильям", "Генри", "Ричард", "Роберт", "Эдуард", "Джон", "Томас", "Джеффри",
            "Роджер", "Хью", "Саймон", "Уолтер", "Гилберт", "Реджинальд", "Балдуин",
            "Алан", "Брайан", "Филип", "Джеймс", "Дэвид"
        ],
        "female": [
            "Элеонора", "Матильда", "Алиенора", "Агнес", "Изабелла", "Беатрис",
            "Сибилла", "Маргарет", "Кэтрин", "Джоан", "Алиса", "Сесилия",
            "Ида", "Мария", "Клеменция", "Амелия", "Филиппа", "Гвендолин", "Аделаида", "Энн"
        ],
        "surnames": [
            "Нормандский", "Анжуйский", "Ланкастерский", "Йоркский", "Плантагенет",
            "Уэссекский", "Глостерский", "Кентский", "Суффолкский", "Ричмондский",
            "Пемброк", "Мортимер", "Грей", "Стюарт", "Говард"
        ],
        "titles": ["Король", "Герцог", "Граф", "Барон", "Рыцарь", "Лорд", "Епископ", "Маркиз"]
    },
    "greek": {
        "male": [
            "Александр", "Аристотель", "Демокрит", "Эпикур", "Гераклит", "Платон",
            "Сократ", "Пифагор", "Гомер", "Софокл", "Еврипид", "Эсхил",
            "Перикл", "Леонид", "Фемистокл", "Агамемнон", "Ахилл", "Одиссей",
            "Гектор", "Тесей"
        ],
        "female": [
            "Афина", "Гера", "Афродита", "Артемида", "Деметра", "Персефона",
            "Гестия", "Ника", "Ирида", "Афродита", "Кассандра", "Андромеда",
            "Елена", "Пенелопа", "Антигона", "Электра", "Ифигения", "Медея",
            "Сапфо", "Гипполита"
        ],
        "surnames": [
            "Афинский", "Спартанский", "Коринфский", "Фиванский", "Македонский",
            "Фессалийский", "Критский", "Родосский", "Милетский", "Эфесский"
        ],
        "titles": ["Царь", "Тиран", "Архонт", "Стратег", "Философ", "Поэт", "Воин", "Герой"]
    },
    "slavic": {
        "male": [
            "Владимир", "Святослав", "Ярослав", "Всеволод", "Изяслав", "Мстислав",
            "Олег", "Игорь", "Святополк", "Вячеслав", "Ярополк", "Глеб",
            "Борис", "Андрей", "Александр", "Дмитрий", "Михаил", "Пётр",
            "Иван", "Василий"
        ],
        "female": [
            "Ольга", "Ярославна", "Владислава", "Людмила", "Мирослава", "Светлана",
            "Евдокия", "Борислава", "Велеслава", "Добромира", "Вера", "Надежда",
            "Любовь", "Мария", "Анна", "Екатерина", "Анастасия", "Татьяна",
            "Елена", "Ирина"
        ],
        "surnames": [
            "Рюрикович", "Владимирович", "Святославич", "Ярославич", "Всеволодович",
            "Изяславич", "Мстиславич", "Олегович", "Игоревич", "Святополкович"
        ],
        "titles": ["Князь", "Воевода", "Боярин", "Дружинник", "Мудрый", "Грозный", "Святой", "Великий"]
    },
    "celtic": {
        "male": [
            "Айдан", "Брендан", "Коннор", "Деклан", "Эйлин", "Финн", "Гэвин",
            "Иан", "Киран", "Лайам", "Мэлколм", "Ниалл", "Она", "Патрик",
            "Рори", "Шон", "Тиадг", "Улисс", "Эмон", "Брайан"
        ],
        "female": [
            "Айслин", "Бриджит", "Кэтлин", "Диана", "Эйлиш", "Финола", "Грэйн",
            "Иона", "Кира", "Лианна", "Майв", "Ниам", "Она", "Рона",
            "Сиана", "Тара", "Уна", "Фиона", "Сиобан", "Морриган"
        ],
        "surnames": [
            "МакКауд", "О'Брайен", "О'Салливан", "МакКарти", "О'Доннелл",
            "МакДонах", "О'Нейл", "МакГрат", "О'Ши", "МакМахон",
            "О'Коннор", "МакДауэлл", "О'Киф", "МакГиннис", "О'Флаэрти"
        ],
        "titles": ["Король", "Вождь", "Друид", "Воин", "Мудрец", "Певец", "Охотник", "Мастер"]
    }
}

class HistoricalNameGenerator:
    def __init__(self, culture="viking", gender="male", count=1,
                 surname=False, title=False, seed=None, color=False):
        self.culture = culture
        self.gender = gender
        self.count = count
        self.surname = surname
        self.title = title
        self.color = color and sys.stdout.isatty()
        if seed is not None:
            random.seed(seed)
        self.culture_data = NAMES.get(culture, NAMES["viking"])

    def generate_name(self):
        names = self.culture_data.get(self.gender, self.culture_data["male"])
        name = random.choice(names)
        result = name
        if self.surname:
            surnames = self.culture_data.get("surnames", [])
            if surnames:
                surname = random.choice(surnames)
                result += f" {surname}"
        if self.title:
            titles = self.culture_data.get("titles", [])
            if titles:
                title = random.choice(titles)
                result += f" {title}"
        return result

    def generate(self):
        return [self.generate_name() for _ in range(self.count)]

    def print_names(self, names):
        culture_names = {
            "viking": "викинг",
            "roman": "римское",
            "medieval": "средневековое",
            "greek": "греческое",
            "slavic": "славянское",
            "celtic": "кельтское"
        }
        culture_label = culture_names.get(self.culture, self.culture)
        gender_label = "мужские" if self.gender == "male" else "женские"
        if self.color:
            print(Fore.CYAN + f"🏛️ Исторические имена ({culture_label}, {gender_label}):")
            for i, name in enumerate(names, 1):
                print(Fore.GREEN + f"{i}. {name}" + Style.RESET_ALL)
        else:
            print(f"Исторические имена ({culture_label}, {gender_label}):")
            for i, name in enumerate(names, 1):
                print(f"{i}. {name}")

    def export_json(self, names, filename):
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump({"culture": self.culture, "gender": self.gender, "names": names}, f, ensure_ascii=False, indent=2)

    def export_csv(self, names, filename):
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(["name"])
            for name in names:
                writer.writerow([name])

    def export_text(self, names, filename):
        with open(filename, 'w', encoding='utf-8') as f:
            f.write("\n".join(names))

def main():
    parser = argparse.ArgumentParser(description="Генератор исторических имён")
    parser.add_argument("--culture", choices=["viking", "roman", "medieval", "greek", "slavic", "celtic"],
                        default="viking", help="Историческая культура")
    parser.add_argument("--gender", choices=["male", "female"], default="male", help="Пол")
    parser.add_argument("--count", type=int, default=1, help="Количество имён")
    parser.add_argument("--surname", action="store_true", help="Добавить фамилию")
    parser.add_argument("--title", action="store_true", help="Добавить титул")
    parser.add_argument("--seed", type=int, help="Seed для генерации")
    parser.add_argument("--output", help="Файл для сохранения")
    parser.add_argument("--format", choices=["text", "json", "csv"], help="Формат вывода")
    parser.add_argument("--color", action="store_true", help="Принудительный цветной вывод")
    args = parser.parse_args()

    gen = HistoricalNameGenerator(
        culture=args.culture,
        gender=args.gender,
        count=args.count,
        surname=args.surname,
        title=args.title,
        seed=args.seed,
        color=args.color
    )
    names = gen.generate()

    if args.output:
        fmt = args.format
        if fmt is None:
            ext = args.output.split('.')[-1].lower()
            if ext == "json":
                fmt = "json"
            elif ext == "csv":
                fmt = "csv"
            else:
                fmt = "text"
        if fmt == "json":
            gen.export_json(names, args.output)
        elif fmt == "csv":
            gen.export_csv(names, args.output)
        else:
            gen.export_text(names, args.output)
        print(f"Результат сохранён в {args.output}")
    else:
        gen.print_names(names)

if __name__ == "__main__":
    main()
